### DISCUSSION TITLE
Implementing hitsound custom samplesets in Lazer

### DISCUSSION THREAD

**Discussion by kstefanowicz** [ppy/osu#29280] [Ideas]
Currently in Lazer, new timing points get created with the legacy attribute `sampleIndex` set to 0. This causes custom hitsounds to not persist to the map. (See #29028)

`sampleIndex` is the [.osu file representation](https://osu.ppy.sh/wiki/en/Client/File_formats/osu_%28file_format%29#timing-points) of the Custom Sampleset menu in Stable:
![image](https://github.com/user-attachments/assets/cf691a9c-f31e-4b7a-bf34-ce2300c0f1f1)

The existing legacy decoder/encoder does support this attribute when importing beatmaps made in Stable, and creates those timing points with the legacy attribute set via the `Suffix` member of `HitSampleInfo`. This allows legacy maps' custom hitsounds to function properly.

For implementing this attribute in Lazer, I see three options:

- Replicate Stable implementation, and add the attribute as a property of timing points
- Replicate Lazer's slider speed and hitsound nodes, and add the attribute as an additional option in `SamplePointPiece`
- Have Lazer create all timing points with `sampleIndex` of 1, and don't bother with values above that

Personally, I like the third option. From what I know the primary use case of the custom samplesets in Stable was to get more than 12 custom hitsounds into a map. With Lazer's implementation of custom sample bank names, that's no longer an issue, and I don't know how many mappers would miss the functionality outside of that use case.

**Comment by cl8n** [comment]
samples, banks, and behavior of custom sample indexes are as much a skinning feature as a mapping one, and I think that consideration is kind of missing from your open PRs regarding forcing sampleIndex=1 (#29311) and saving custom bank names (#29566).

in stable (and lazer gameplay afaik):
- `S` plays the skin's `soft-<type>.wav`
- `S:C1` plays the map's `soft-<type>.wav`, with fallback to the skin's `soft-<type>.wav`
- `S:C2` and onward play the map's `soft-<type><index>.wav`, with fallback to the skin's `soft-<type>.wav`

I agree with what you've implied about the first case being unnecessary. if the mapper wants to strictly use the skin's samples, then they can pick `S:C1` and just not include the respective files in their map. so I think for non-legacy maps removing "sample index 0" and *defaulting* (not *forcing*) sample index to 1 would be a fine simplification of this system.

I don't agree with your further changes as-is that rely on custom bank names though. currently a `<custom-bank>-<type>.wav` sample has no clear fallback bank to use for skin lookups, and it's probably not going to sound good if you force that fallback bank to be any particular default. additionally, there's no way to get this to function in stable without changes due to the new custom filenames just not being supported there.

my suggestions for how to implement this feature instead: (crossed out 2 and 3 because olibomby's suggestion is overall better imo)

1. the stable-ish way: remove "custom banks" and keep the concept of sample indexes alive, but move their editor functionality to individual samples rather than timing points where possible, like how SV changes did it.

2. ~~if custom banks are nice to have in the editor: do the above but then also implement some kind of aliasing system so that you can refer to `N:C2` as a more meaningful `My Custom Bank`. the filename would still be `normal-<type>2.wav`, though~~

3. ~~if custom banks are considered much better overall *and making some changes to stable is acceptable*:~~
   - ~~give custom banks the additional option of which default bank they fallback to~~
   - ~~save fallback banks in the custom bank section of legacy format you've already created in #29566~~
   - ~~implement appropriate fallback behavior in both lazer and stable (map's `<custom-bank>-<type>.wav` --> skin's `<fallback-bank>-<type>.wav`)~~

**Reply by OliBomby** [reply]
If we're going into changing how banks work, I have a fourth suggestion for how to implement custom hitsounds:

4. Let each sample define fallback bank and hitsound for skin's hitsound when custom disabled. Then add an optional custom hitsound filename property that if defined (and beatmap customs enabled) looks for the filename in the beatmap skin for a custom sample sound.
I feel ilke this would be the simplest implementation of custom hisounds and it matches how I implemented it in my Hitsound Studio tool.

**Reply by cl8n** [reply]
that sounds cool too for a lot of cases, but I can't tell if it would entirely replace the need for either custom samplesets or custom banks. for example I might have a whole section of a map where I want to use my N:C2 (...or custom bank, if it gets implemented like that) hitsounds and it seems inconvenient if I have to select custom hitsound files for that whole section instead of just adding clap or whistle or finish like usual. I also feel like this would complicate hitnormal, since that is assumed to always play for the given sampleset.

I guess it all depends on how this feature is presented, maybe you have more clever ideas in mind to avoid these UX problems

**Reply by OliBomby** [reply]
Yeah it does pose some UX problems for the case where you use custom hitsounds but not that much that you're changing custom bank every object. The implementation in stable seems especially optimised for this specific use case.
I'm not sure what you mean with that it complicates the hitnormal, it would work just like the other samples.

I think all you need to solve the UX is additional tools to quickly assign the custom samples, both while mapping and in post to make bulk changes.
A good addition would be to automatically assign a default custom hitsound filename whenever you add a hitsound, so add `drum-hitwhistle` as the custom filename when you add a drum-hitwhistle etc. Also add an index number so you can have multiple presets. This will only affect placement of hitsounds and it will be fully customizable afterwards.
I dislike how non-descript the filenames are in stable (`soft-hitclap2.wav` doesn't tell you what sound it is), so it would be nice to also have a menu to customize these defaults. Basically a mapping where you can tell it to assign for example `bass snare.wav` instead of `soft-hitclap2`. This will make the filenames more descriptive and eliminates the need to duplicate sample files between custom sample sets.

**Reply by cl8n** [reply]
that does sound like a good direction then. compared to stable it seems more useful *and* less convoluted if presented right. I suppose it can also be legacy-exportable if that process does some extra work to map everything back into custom samplesets and rename/copy files accordingly, so probably no need to update stable

I updated my post with my new opinion about these options -- yours is definitely best, but I still think 1 would be a valid choice right now if the goal is to quickly and more easily get this feature in parity with stable.

**Reply by kstefanowicz** [reply]
This is how I imagine the UX (pardon the MS Paint):

![image](https://github.com/user-attachments/assets/1452e0bf-ee5b-4929-b710-209ac47d709e)

Each "▶" would play the current-assigned hitsound for that combination. Each "+" would open a pop-up to select an existing audio file added to the map or drag in a new audio file.

Assuming custom sound banks are a thing, users would click the "+" at the bottom of the first column to add a new custom bank.

This UX doesn't address the fallback options though, I don't know what the cleanest way to present that would be.


> I dislike how non-descript the filenames are in stable (soft-hitclap2.wav doesn't tell you what sound it is)

To my knowledge the game chooses a hitsound based entirely on the filename. To have the filename be separate, I guess it would need to either copy the file and rename it to fit, or change how audio files are chosen?

**Comment by bdach** [comment]
If I understand the conversation here correctly, and how custom sample banks work, at current I'm split between the following two concepts (hopefully my handwriting isn't completely illegible):

<img width="1404" height="2008" alt="Quick sheets - page 31" src="https://github.com/user-attachments/assets/9c39114b-b5f3-4bdd-b2ea-af5159afba9b" />

At this time I'm not necessarily sure which is better. It probably depends on how hard people are bought into the current stable hit sounding workflow. Option 1 is more or less directly that, but less horrible UI compared to stable; option 2 is what @OliBomby is advocating for, but in this case mappers will have to relinquish control over how sample bank indices are assigned and I'm not necessarily sure they are prepared for this.

Will put some feelers out in mapping spheres I have access to and see whether I get any usable feedback. If anyone else following this thread wants to do the same, then by all means. @peppy gonna put this in your orbit too, in case you have some immediate vetoes or other general feedback on the above.

**Reply by bdach** [reply]
And to preempt criticism that this isn't the "DAW-like" "hitsound timeline" view people probably want to see - I don't want to commit to that sort of scope right now. My primary goal is to unblock interacting with sample sets which is a major pain for people using lazer editor right now, from feedback I've received via the mappers guild quest that's ongoing (been ongoing?)

**Reply by OliBomby** [reply]
Converting option 2's samples to stable's custom banks is definitely possible with Hitsound Studio's algorithm. I can implement this in the legacy exporter if we ever go that route. That said option 1 is probably the best if we want to get something out quick because its more compatible with the current file format. Having good GUI to make custom sample banks more intuitive is already a big improvement over stable.

I think when the file format evolves more and allows more freedom, the custom sample banks could be interpreted as user-defined presets for the more granular configurations as in option 2, so option 1 and 2 (and potential DAW timeline thing) could be combined in the future.

Edit: Actually if we ignore the technical debt of stable compatibility, option 1 feels terrible and convoluted. I think we should strive to make it as straight forward as possible for new mappers. I want the UX to be 1-to-1 with the data model that we use in lazer internally which is `List<HitSampleInfo>`, so when a mapper is deciding what sounds to put on a hit, they can just write down a list of samples instead of having to think about 'whistles' or 'finishes'.

**Reply by river-is-mmm** [reply]
Personally, even in stable, I tend to strive to either limit my hitsounds to one sampleset (sometimes even just default) or at least avoid having to change samplesets any more often than section-by-section. Fine-grain sampleset index management is super tedious; I view option 2 less as "relinquishing control" and more as just not having to worry about it anymore. I'd be _fine_ with option 1 since it's basically a glow-up for what I'm already used to, but something along the lines of option 2 would be a welcome change for me (as long as the design avoids tedium in cases where I'd mostly be picking the same sample every time anyway).

> when a mapper is deciding what sounds to put on a hit, they can just write down a list of samples instead of having to think about 'whistles' or 'finishes'

I'm kinda torn on this. On the one hand, _yes_. I want to just pick whatever sounds I want from a list of samples with descriptive names, that'd be awesome. On the other hand, the whistle/finish/clap paradigm is so heavily integrated into osu! at this point that deliberately managing _that_ does have some value; notably, there's a ranking criterion about it:

> **Avoid replacing the hit finish in soft/normal samplesets with frequently used custom hitsound samples.**

Even beyond that, making sure that my W/F/C placements sound good in general even when someone's skin replaces them with their own distinct samples is also something I do consider when hitsounding (although I tend to compromise on that for the sake of convenience right now).

Putting all that in writing helped an idea crystallise in my mind: A system where you assign each of your custom samples to be a hitnormal, whistle, finish, or clap (as well as the choice of normal/soft/drum I guess). Then, when placing samples on objects, you've got a list of samples to choose from for each hitsound "slot". It'd also be helpful to be able to select primary samples for the slots so that our existing hitsounding hotkeys are still useful and convenient. I feel like this could be a decent melding of Oli's ideal and stable compatibility.

**Reply by bdach** [reply]
> A system where you assign each of your custom samples to be a hitnormal, whistle, finish, or clap (as well as the choice of normal/soft/drum I guess). Then, when placing samples on objects, you've got a list of samples to choose from for each hitsound "slot".

I didn't spell this out on my crude drawings but I did have it in mind that this was something that could be done as an extension of option 2.

**Reply by bdach** [reply]
For general awareness, I am going to try executing option 1. - the stable-like UI.

This may seem to be dismissive of the voices that say stable's hitsound system sucks and shouldn't come back. The primary reason why I am trying the stable system first are two-fold:

1. To cut scope. I already foresee significant roadblocks in making even the stable-like option work as I think it should. Adding more complexity on top to make any new system work is too much scope for me to manage short-term - and I am pretty confident that it will be more complexity on top rather than a parallel path requiring comparable effort.
2. For the foreseeable future stable has to exist, and so any new system has to be backwards-compatible with stable. While yes, I can see how you make conversions so that a new system can be backported to stable, I'm not sure this is in the best interests of users, BNs/NATs, etc. that will have to interface with the results of said conversions. I have reservations about introducing further hidden systems to areas where users are used to exercising significant degree of control. Not to mention potential effects on things like the ranking criteria or some such.

Think of it this way:
- I am making a proof of concept first. The results of that effort don't necessarily mean anything. I just have to start *somewhere* to implement this and that's where I'm deciding to start.
- None of this is likely final anyway. As touched on earlier, hitsounders desire a DAW-like interface to interact with samples anyway. I am interested in a short-term measure to implement an important feature that makes the lazer editor not on par with stable.

Thanks for understanding.

**Comment by Serizawa-Haruki** [comment]
I'd like to mention this post I made some time ago since it's relevant to the current discussion: https://github.com/ppy/osu/discussions/32725

In short, lazer shouldn't adopt the flawed hitsound sysrem from stable.

**Reply by bdach** [reply]
Stable is still relevant. Anything I make *short term* must work on stable. Therefore I do not foresee any major overhauls of how hitsounds work, until stable becomes less relevant to a degree that we can start considering making some clean breaks from it.

**Reply by Serizawa-Haruki** [reply]
Yes stable is obviously still relevant, especially for mapping, but I never said the new system shouldn't be compatible with it. How that would work is a different story, but either way it shouldn't be a reason to recreate a system with the same flaws. It's similar to the classic mod, just because it always used to be  this way doesn't mean it can't/shouldn't be changed.

**Comment by bdach** [comment]
Here's an early preview of the ongoing proof of concept of the 'stable-like' system recreation:

https://github.com/user-attachments/assets/e60b9d51-63fd-425a-b165-f43592c8ccdd

What's missing is a coherent way to manage the sample sets (likely to be added to setup screen). The reason it's not in the popover as it was on my drafts is that (1) I realised it was likely going to need popovers-in-popovers and (2) the stable system works for the entire set, as in doing *any* change to hitsounds impacts the entire map, so it seems more suited there.

(Note that the above concerns *also would apply* if I went the "freeform sample selection" route. Imagine being able to combine the stable way of doing things *with* the new system. And also imagine what happens when someone tries to do a hitsounded guest difficulty, etc. etc. These are other reasons why I want to start with the stable way and see where all of the minefields are.)

Code available for preview at https://github.com/ppy/osu/compare/master...bdach:osu:custom-sample-bank-support?expand=1.
